### DISCUSSION TITLE
How does SCons inject symbols into global scope?

### DISCUSSION THREAD

**Discussion by abitrolly** [SCons/scons#4200] [Q&A]
How does SCons make global symbols available without imports?

```python
$ cat SConstruct 
print(BoolVariable)
```
```python
$ scons         
scons: Reading SConscript files ...
<function BoolVariable at 0x7f5c57846f80>
scons: done reading SConscript files.
scons: Building targets ...
scons: `.' is up to date.
scons: done building targets.
```

Is it possible to inject symbols dynamically? Like when a custom tool is loaded.

**Comment by bdbaddog** [comment]
A custom tool doesn't inject symbols into the global symbols.
It add's itself into the provided Environment()

There's not a SCons provided way to add to the globals() in the context of a SConscript/SConstruct.

You can `Import("*")` which might solve your issue depending on what you're trying to do.
Perhaps explain your use model and we can provide a workable solution.

**Reply by abitrolly** [reply]
I am just interested to know how SCons works. How different parts of it can be augmented without committing them to the core. Might reuse the same approach for my own Python scripts.

> You can `Import("*")` which might solve your issue depending on what you're trying to do.

Is that it? https://scons.org/doc/latest/HTML/scons-api/_modules/SCons/Script/SConscript/#SConsEnvironment.Import Not sure what it does.

>  Perhaps explain your use model and we can provide a workable solution.

I don't need a workable solution, but the use model I'd like to see is this. Given a custom dir for SCons files, like `.scons`, I want to drop my `Sophisticated` variable into `.scons/Variables/Sophisticated.py` so that it became available to my scripts and I can use it the same way `BoolVariable` is used. If SCons doesn't work this way, I am interested to know how to get it there.

**Comment by bdbaddog** [comment]
Add it in `.site_scons/__init__.py` 
Have you read the users guide or manpage thoroughly?

Search for site_scons in the manpage, and look at : https://scons.org/doc/production/HTML/scons-user.html#idm46358254405712

**Reply by abitrolly** [reply]
Yes, I am aware of `site_scons/site_tools`, and "where to put tools" is not what I am asking.

I'd like to understand how/where SCons injects symbols in the global scope of `SConstruct`? Is it a static or dynamic process? If process is static, then the injection is probably done by Python itself at import time before `SConstruct` is read. If dynamic, then it must be possible to inject new symbols after `SConstruct` started to execute.

**Comment by mwichmann** [comment]
As far as the implementation, that "magic" happens in `SCons/Script`. The sconscript files are successively exec'd using the existing globals after some stack fiddling.

**Reply by abitrolly** [reply]
Found only one `exec` here.

https://github.com/SCons/scons/blob/bc959a986e9998b32c2a74120794a38c99541822/SCons/Script/__init__.py#L379

It doesn't looks like it executes `SConscript` files. Rather sets global variables in itself.

**Comment by bdbaddog** [comment]
I wasn't talking about `site_scons/site_tools` I was talking about `site_scons/site_init.py`
Any symbols set in there should end up in the global name space.
Which the docs explain... 

```
Before reading the SConscript files, scons looks for a directory named site_scons in various system directories and in the directory containing the SConstruct file or, if specified, the directory from the [--site-dir](https://scons.org/doc/production/HTML/scons-man.html#opt-site-dir) option instead, and prepends the ones it finds to the Python module search path (sys.path), thus allowing modules in such directories to be imported in the normal Python way in SConscript files. For each found site directory, (1) if it contains a file site_init.py that file is evaluated, and (2) if it contains a directory site_tools the path to that directory is prepended to the default toolpath. See the [--site-dir](https://scons.org/doc/production/HTML/scons-man.html#opt-site-dir) and [--no-site-dir](https://scons.org/doc/production/HTML/scons-man.html#opt-no-site-dir) options for details on default paths and controlling the site directories.
```

**Reply by abitrolly** [reply]
Missed that one. Good to know. That still doesn't explain the underlying mechanism how symbols are injected.

**Reply by bdbaddog** [reply]
The source code is there for all to read. :)

**Reply by abitrolly** [reply]
Too much code. An entrypoint would be welcome. )

**Reply by bdbaddog** [reply]
That's what grep's for!

`SCons/Script/SConscript.py`
Beyond that you're on your own.

**Reply by bdbaddog** [reply]
`SConscript()` calls end up building a set of globals based on previous globals. This happens at each SConscript call. So parent SConscript calls would have different globals from children.  It's a bit complicated, and likely this implementation is not that useful in other contexts.

Python's `eval()` can be passed a set of globals. That's the basis for SConscript/SConstruct processing.

**Reply by abitrolly** [reply]
Looks like this is it.

https://github.com/SCons/scons/blob/bc959a986e9998b32c2a74120794a38c99541822/SCons/Script/SConscript.py#L196

And here is the first frame in stack where globals are supplied by `BuildDefaultGlobals()`.

https://github.com/SCons/scons/blob/bc959a986e9998b32c2a74120794a38c99541822/SCons/Script/SConscript.py#L108

For `SConscript` that is somehow piped to `stdin`, if I read this correctly.

1. First a `Frame` object is created on `call_stack`, with `BuildDefaultGlobals()`
2. Then `SConscript` is executed passing it globals from `Frame`
3. Then the `Frame` is popped of the stack.

So if no `globals` are shared once script starts execution, then `SCons` can not dynamically insert new symbols after `exec` started.
